### PR TITLE
[Bugfix] Fix LongRoPE bug

### LINKED ISSUE #6438: [Bug]: Severe computation errors when batching request for microsoft/Phi-3-mini-128k-instruct
### Your current environment

I'm not able to run `collect_env.py` on this workstation

vllm == 0.5.1
vllm-flash-attn == 2.5.9
torch == 2.3.0

Tested on a single A100-80GB

The following message was observed:
```
Cannot use flash attention-2 backend due to sliding window
Using XFormers backend
```

### 🐛 Describe the bug

Issue:
- Scenario 1: calling `LLM.generate` with `max_num_seqs=64`.
- Scenario 2: sending requests to Vllm OpenAI compatible server with 50 concurrent calls
For both scenarios, many responses do not make any sense. While some responses seemed perfectly reasonable, others contain an excessive amount of white spaces and repeated characters.

I was not able to reproduce this using single calls in Scenario 2. With a single call to the sever, all responses seemed correct.

My speculation is that the batching computation for Phi-3 is incorrect.

### PR DESCRIPTION
**Pull request by garg-amit**
This PR addresses [this issue](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/discussions/85) by using the number of original input tokens to decide using short or long factors to avoid the switch happening on one generation request: if the number of original input tokens < original_max_position_embeddings, it uses short factors; other than that, it uses long factors.

FIX #6438

Models tested before this PR:

- microsoft/Phi-3.5-mini-instruct
- microsoft/Phi-3-mini-4k-instruct
- microsoft/Phi-3-small-8k-instruct
- microsoft/Phi-3-small-128k-instruct
- microsoft/Phi-3.5-MoE-instruct
- OpenVINO/Phi-3-mini-128k-instruct-int4-ov

Backends tested:

- XFORMERS
- FLASHINFER
- FLASH_ATTN
- OPENVINO
- ROCM_FLASH

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

@robertgshaw2-neuralmagic @mgoin

### CODE REVIEW

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic yes, the PR is ready for review now

**Comment by robertgshaw2-neuralmagic** [comment]
@garg-amit what would need to happen to trigger this feature? Is there anything inside LM Eval harness that we could use to trigger and confirm this is working correctly in our CI

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic This fix addresses the scenario where the input length is shorter than the original_max_position_embeddings, but the combined input and output exceed the original_max_position_embeddings. I’m not sure which evals might trigger this case, but to check parity with other scenarios: for short-context evals, MMLU and GSM8k, and for long-context evals, Qasper can be triggered.

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic we performed some local testing to trigger the above scenario with Phi-3.5-mini-instruct. Here are the outputs before and after the fix.

`Prompt: `

```
Microsoft Corporation is an American multinational corporation and technology company headquartered in Redmond, Washington.[2] Its best-known software products are the Windows line of operating systems, the Microsoft 365 suite of productivity applications, the Azure cloud computing platform and the Edge web browser. Its flagship hardware products are the Xbox video game consoles and the Microsoft Surface lineup of touchscreen personal computers. Microsoft ranked No. 14 in the 2022 Fortune 500 rankings of the largest United States corporations by total revenue;[3] and it was the world's largest software maker by revenue in 2022 according to Forbes Global 2000. It is considered one of the Big Five American information technology companies, alongside Alphabet (parent company of Google), Amazon, Apple, and Meta (parent company of Facebook).
...
[62] The European Union imposed another fine of €899 million ($1.4 billion) for Microsoft's lack of compliance with the March 2004 judgment on February 27, 2008, saying that the company charged rivals unreasonable prices for key information about its workgroup and backoffice servers.[63] Microsoft stated that it was in compliance and that "these fines are about the past issues that have been resolved".[64] 2007 also saw the creation of a multi-core unit at Microsoft, following the steps of server companies such as Sun and IBM.

Gates retired
```

`Prompt length: 4093`

Output before the fix:

```
 as chief software architectingessis.
as.




00orx








0.0omist.0aringing, with.
.
.





...


of0X.

game,


of000000000000s.0000000000000000000.
0X


s
0000X00XX



X0s of
of
of0orofof.0000xxon, of000xx.,
000


time,00o
XXXor0s.
0000XXXXX00

sXor.or.0s000s.
of.0on.
of.





000s
000oXNT.

00



such.
onon.

XorXXs00

```

After the fix:

```
 as chief software architect in June 2006, and as chairman of the board in February 2008.[65] In 2008, Microsoft released Windows Phone 7, a new mobile operating system for smartphones, and the Xbox 360 S, a revised version of the Xbox 360.[66][67] The company also released the Zune HD, a digital media player with video playback, and the Surface, a tablet computer.[68]

In 2009, Microsoft released Windows 7, a new version of Windows that was generally well received.[69] The company also released the Xbox 360 Wireless Gaming Receiver, a wireless adapter for the Xbox 360.[70]

In 2010, Microsoft released the Kinect, a motion-sensing input device for the Xbox 360 and Xbox One.[71] The company also released the Surface 2, a tablet computer with a new design and improved hardware.[72]

In 2011, Microsoft released Windows 8, a new version of Windows that was generally well received.[73] The company also released the Xbox 360 E, a revised version of the Xbox 360.[74]
...
Microsoft released

```

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic can I please get a review?

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic can I please get a review?

Yes. Will get this in for v0.6.2

**Comment by robertgshaw2-neuralmagic** [comment]
> > @robertgshaw2-neuralmagic can I please get a review?
> 
> Yes. Will get this in for v0.6.2

@garg-amit - I was waiting for tests to be green to merge. Are you planning to address them?

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic I’ve fixed most of the test cases. The remaining test cases seem unrelated to the current PR. Please let me know if there is anything else that needs to be fixed.

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic I’ve fixed most of the test cases. The remaining test cases seem unrelated to the current PR. Please let me know if there is anything else that needs to be fixed.

Thanks. Will review and merge tomorrow

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -177,7 +179,16 @@ def forward(
     ) -> torch.Tensor:
         qkv, _ = self.qkv_proj(hidden_states)
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
-        q, k = self.rotary_emb(positions, q, k)
+        if self.rope_scaling is None or not isinstance(
+                self.rotary_emb, Phi3LongRoPEScaledRotaryEmbedding):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/llama.py:183]: Why do we need to add this special case to `llama.py` when we subclass `LlamaAttention` in `phi3.py`?

File: vllm/attention/backends/abstract.py [hunk 2]
```diff
@@ -105,6 +105,10 @@ class AttentionMetadata:
     # in block 0, and 1st slot in block 1, respectively.
     slot_mapping: torch.Tensor
 
+    # Number of original input tokens (without any decoding).
+    # Some model (phi3-) need this info to decide model settings
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/backends/abstract.py:109]: Can you improve the comment here to specify this is for phi3 LongRoPE

File: vllm/attention/backends/abstract.py [hunk 3]
```diff
@@ -105,6 +105,10 @@ class AttentionMetadata:
...
+    num_orig_input_tokens_tensor: torch.Tensor
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/backends/abstract.py:110]: Is this actually `Optional[torch.Tensor]`?

**Comment by robertgshaw2-neuralmagic** [comment]
AMD CI has a broken node. Okay to ignore the issues

File: vllm/attention/backends/utils.py [hunk 4]
```diff
@@ -188,7 +188,8 @@ def _add_seq_group(
                                  self.block_size, inter_data.block_tables)
 
     def build(self, seq_lens: List[int], query_lens: List[int],
-              cuda_graph_pad_size: int, batch_size: int):
+              num_orig_input_tokens_list: List[int], cuda_graph_pad_size: int,
+              batch_size: int):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/backends/utils.py:192]: please update the docstring?

File: vllm/attention/backends/utils.py [hunk 5]
```diff
@@ -256,13 +257,18 @@ def build(self, seq_lens: List[int], query_lens: List[int],
                      dtype=query_start_loc.dtype,
                      out=query_start_loc[1:])
 
+        num_orig_input_tokens_tensor = torch.tensor(num_orig_input_tokens_list,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/backends/utils.py:260]: Should this use the `async_tensor_h2d` API?

File: vllm/worker/model_runner.py [hunk 6]
```diff
@@ -214,6 +215,8 @@ def __init__(
             input_positions: Optional[List[List[int]]] = None,
             mrope_input_positions: Optional[List[List[List[int]]]] = None,
 
+            # The number of original input tokens of each sequence
+            num_orig_input_tokens_list: Optional[List[List[int]]] = None,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/worker/model_runner.py:219]: Can you explain / comment why this is a List of Lists? I understand the outer list is the sequence id, but what is the inner list for?

The other model runners are a single List, which is more intuitive to me

File: vllm/worker/model_runner.py [hunk 7]
```diff
@@ -503,6 +515,13 @@ def _compute_lens(self, inter_data: InterDataForSeqGroup, seq_idx: int,
             inter_data.input_positions[seq_idx].extend(
                 range(context_len, seq_len))
 
+        if (seq_len - context_len) == 1:
+            inter_data.num_orig_input_tokens_list[seq_idx].append(
+                seq_data.get_prompt_len())
+        else:
+            inter_data.num_orig_input_tokens_list[seq_idx].extend(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/worker/model_runner.py:522]: Can you explain / comment why the `context_len` matters here?

File: vllm/model_executor/models/llama.py (hunk 1 above)
Comment (garg-amit) [reply vllm/model_executor/models/llama.py:183]: To handle cases for models that inherits Llama model, such as Phi-3-vision, and use Phi3LongRoPEScaledRotaryEmbedding.

File: vllm/attention/backends/abstract.py (hunk 2 above)
Comment (garg-amit) [reply vllm/attention/backends/abstract.py]: Updated

File: vllm/attention/backends/utils.py (hunk 4 above)
Comment (garg-amit) [reply vllm/attention/backends/utils.py:192]: Updated

File: vllm/attention/backends/utils.py (hunk 5 above)
Comment (garg-amit) [reply vllm/attention/backends/utils.py]: Fixed

File: vllm/worker/model_runner.py (hunk 6 above)
Comment (garg-amit) [reply vllm/worker/model_runner.py:219]: @robertgshaw2-neuralmagic the goal was to ensure that `num_orig_input_tokens_list` has the same shape as `input_positions`. This would simplify switching the logic based on `num_orig_input_tokens_tensor` instead of the `positions` tensor within the `Phi3LongRoPEScaledRotaryEmbedding` class.

File: vllm/worker/model_runner.py (hunk 7 above)
Comment (garg-amit) [reply vllm/worker/model_runner.py:522]: Same as above (to keep the shape same as the `input_positions` tensor)

File: vllm/attention/backends/abstract.py (hunk 3 above)
Comment (garg-amit) [reply vllm/attention/backends/abstract.py]: @robertgshaw2-neuralmagic thanks for the suggestion! Fixed it

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic The PR is ready for another look.

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic Please let me know if there is anything else that needs to be addressed.

**Comment by robertgshaw2-neuralmagic** [comment]
Was OOO last week. My apologies. Will review today

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic ping

**Comment by garg-amit** [comment]
@zhuohan123 @youkaichao @WoosukKwon could you help review this PR?

**Comment by robertgshaw2-neuralmagic** [comment]
Im so sorry @garg-amit - this slipped off my radar

**Comment by robertgshaw2-neuralmagic** [comment]
LGTM! Getting the CI green

**Comment by robertgshaw2-neuralmagic** [comment]
wokring to get it green

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic I’ve fixed the above test case. However, the remaining test cases appear to be unrelated to the current PR. I tried to resolve them, but I couldn’t. Please let me know if those need to be resolved. I would appreciate your help.

**Comment by garg-amit** [comment]
@robertgshaw2-neuralmagic when are you planning to merge this PR?

**Comment by DarshanDeshpande** [comment]
@garg-amit @robertgshaw2-neuralmagic Is there an update on this PR?

**Comment by garg-amit** [comment]
Awaiting Robert's response. @robertgshaw2-neuralmagic, could you please confirm if DCO is a hard requirement? While rebasing, I'm being prompted to sign off on commits that I didn't author. If required, I can create a new PR with the same changes.

**Comment by robertgshaw2-neuralmagic** [comment]
> Awaiting Robert's response. @robertgshaw2-neuralmagic, could you please confirm if DCO is a hard requirement? While rebasing, I'm being prompted to sign off on commits that I didn't author. If required, I can create a new PR with the same changes.

DCO can be overidden. Dont worry about rebasing.

**Comment by garg-amit** [comment]
@robertgshaw2-redhat I've resolved the merge conflicts and addressed the related test cases. However, some tests are still failing with `requests.exceptions.ReadTimeout: (ReadTimeoutError("HTTPSConnectionPool(host='huggingface.co', port=443): Read timed out. (read timeout=10)")`. Could you please review the PR? Additionally, I would appreciate it if you could override the DCO. Thank you!
